File: 프로그래머스/lv3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    MOD = 1000000007
    
    # Create a grid to store the number of paths to each cell
    grid = [[0] * (n+1) for _ in range(m+1)]
    
    # Mark the starting cell with 1 path
    grid[1][1] = 1
    
    # Mark flooded cells as -1 to avoid them
    for puddle in puddles:
        grid[puddle[0]][puddle[1]] = -1
    
    for i in range(1, m+1):
        for j in range(1, n+1):
            # Skip flooded cells and the starting cell
            if grid[i][j] == -1 or (i == 1 and j == 1):
                continue
            
            # Calculate the number of paths to this cell
            # by adding paths from the cell above and the cell to the left
            if grid[i][j-1] != -1:
                grid[i][j] += grid[i][j-1]
            if grid[i-1][j] != -1:
                grid[i][j] += grid[i-1][j]
            
            grid[i][j] %= MOD
    
    return grid[m][n]

File: 프로그래머스/lv3/42898. 등굣길/등굣길.cpp
#include <string>
#include <vector>

using namespace std;

const int MOD = 1'000'000'007;

int solution(int m, int n, vector<vector<int>> puddles) {
    vector<vector<int>> dp(n+1, vector<int>(m+1, 0));

    // 물에 잠긴 지역 표시
    for (auto& puddle : puddles) {
        dp[puddle[1]][puddle[0]] = -1;
    }

    dp[1][1] = 1;  // 시작 지점

    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= m; ++j) {
            if (i == 1 && j == 1) continue;  // 시작 지점은 이미 1로 설정됨
            if (dp[i][j] == -1) {  // 물에 잠긴 지역은 0으로 설정
                dp[i][j] = 0;
                continue;
            }
            dp[i][j] = (dp[i-1][j] + dp[i][j-1]) % MOD;
        }
    }

    return dp[n][m];
}
